src/OpenDebugAD7/Tracepoint.cs

```diff
@@ -0,0 +1,546 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+
+using Microsoft.VisualStudio.Debugger.Interop;
+
+namespace OpenDebugAD7
+{
+    internal class Tracepoint
+    {
+        // LogMessage after it has been Parsed()
+        public string LogMessage { get; set; }
+
+        // Map of the index in LogMessage to insert the evaluated expression ($TOKEN or { expression })
+        private readonly IDictionary<int, string> m_indexToExpressions;
+
+        private Tracepoint(string logMessage)
+        {
+            m_indexToExpressions = new Dictionary<int, string>();
+            LogMessage = Parse(logMessage);
+        }
+
+        internal static Tracepoint CreateTracepoint(string logMessage)
+        {
+            return new Tracepoint(logMessage);
+        }
+
+        internal int GetLogMessage(IDebugThread2 pThread, uint radix, string processName, out string logMessage)
+        {
+            int hr = HRConstants.S_OK;
+            string message = LogMessage;
+
+            // There is strings to interpolate in the log message.
+            if (m_indexToExpressions.Count != 0)
+            {
+                hr = GetInterpolatedLogMessage(message, pThread, radix, processName, out message);
+            }
+
+            logMessage = message;
+            return hr;
+        }
+
+        private int GetInterpolatedLogMessage(string logMessage, IDebugThread2 pThread, uint radix, string processName, out string message)
+        {
+            int hr = HRConstants.S_OK;
+
+            if (pThread == null)
+            {
+                message = string.Format(CultureInfo.InvariantCulture, "Unable to interpolate logMessage because current thread is missing.");
+                return HRConstants.E_FAIL;
+            }
+
+            // Get topFrame
+            IEnumDebugFrameInfo2 frameInfoEnum;
+            hr = pThread.EnumFrameInfo(enum_FRAMEINFO_FLAGS.FIF_FRAME | enum_FRAMEINFO_FLAGS.FIF_FLAGS, Constants.EvaluationRadix, out frameInfoEnum);
+            if (hr < 0)
+            {
+                message = "Unable to interpolate logMessage because frames could not be retrieved.";
```

Comment:
string.format

Reply:
Should your warnings/errors be in a resx in case we decide to do localization in the future?